### PR TITLE
Make undo_stack configurable and add quit function for exception hand…

### PR DESCRIPTION
**Pull request by linhuang-blockone**
…ling

## Change Description

Add `--disable-undo-stack` to `rodoes` startup option to disable undo_stack handling.

Gracefully shut down `rodeos` when a fork happens if undo stack is not enabled.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [x] Other

Manual testing.

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: programs/rodeos/ship_client.hpp [hunk 1]
```diff
@@ -185,6 +187,14 @@ struct connection : connection_base {
    void catch_and_close(F f) {
       try {
          f();
+      } catch (const std::runtime_error& e) {
+         elog("${e}", ("e", e.what()));
+         
+         if (std::strstr(e.what(), "quit rodeos") != nullptr) {
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:193]: Do this a diff way. Can use EOS exceptions like everywhere else in our code.

File: programs/rodeos/ship_client.hpp [hunk 2]
```diff
@@ -33,6 +33,7 @@ struct connection_callbacks {
    virtual bool received(ship::get_blocks_result_v1& result, eosio::input_stream bin) { return true; }
    virtual bool received(ship::get_blocks_result_v2& result, eosio::input_stream bin) { return true; }
    virtual void closed(bool retry) = 0;
+   virtual void quited() = 0;
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:36]: Remove `quited`. `closed` already has a path that calls `appbase::app().quit()`. Add another option to `closed` on if to call `quit`.

File: programs/rodeos/ship_client.hpp [hunk 3]
```diff
@@ -61,6 +62,7 @@ struct connection_base {
    virtual void send(const ship::request& req) = 0;
    virtual void request_blocks(const ship::get_status_result_v0& status, uint32_t start_block_num, const std::vector<ship::block_position>& positions, int flags) = 0;
    virtual void close(bool retry) = 0;
+   virtual void quit() = 0;
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:65]: Not needed. See above.

File: programs/rodeos/ship_client.hpp [hunk 4]
```diff
@@ -215,6 +225,15 @@ struct connection : connection_base {
...
    }
+
+   void quit() {
+      ilog("quit");
+      if (callbacks) {
+         callbacks->quited();
+      } else {
+         wlog("callbacks is not initiated yet when quit is called");
+      }
+   }
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:236]: Not needed. See above.

**Review by heifner — CHANGES_REQUESTED** [review]

File: programs/rodeos/ship_client.hpp (hunk 1 above)
Comment (linhuang-blockone) [reply programs/rodeos/ship_client.hpp]: Thanks for pointing this out. I did not like strstr either and was trying to follow the original throw style. Changed to `eosio::chain::unlinkable_block_exception`.

File: programs/rodeos/ship_client.hpp (hunk 2 above)
Comment (linhuang-blockone) [reply programs/rodeos/ship_client.hpp]: Initially I added a `quitting` parameter to `closed`. Then I was worry that made `closed` do two things and a specific `quitted`  might be cleaner. Changed to use an additional option to `closed`.

File: programs/rodeos/ship_client.hpp (hunk 4 above)
Comment (linhuang-blockone) [reply programs/rodeos/ship_client.hpp]: Removed.

File: programs/rodeos/ship_client.hpp [hunk 5]
```diff
@@ -32,7 +33,7 @@ struct connection_callbacks {
    virtual bool received(ship::get_blocks_result_v0& result, eosio::input_stream bin) { return true; }
    virtual bool received(ship::get_blocks_result_v1& result, eosio::input_stream bin) { return true; }
    virtual bool received(ship::get_blocks_result_v2& result, eosio::input_stream bin) { return true; }
-   virtual void closed(bool retry) = 0;
+   virtual void closed(bool retry, bool quititingi = false) = 0;
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:36]: spelling of quitting.

Comment (heifner) [reply programs/rodeos/ship_client.hpp:36]: I would prefer for `quitting` not to be defaulted.

File: programs/rodeos/ship_client.hpp [hunk 6]
```diff
@@ -208,11 +212,11 @@ struct connection : connection_base {
       } catch (...) { elog("exception while closing"); }
    }
 
-   void close(bool retry) {
-      ilog("closing state-history socket");
+   void close(bool retry, bool quitting = false) {
```
Comment (heifner) [inline programs/rodeos/ship_client.hpp:215]: Would prefer for `quitting` not to be defaulted.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/rodeos/rodeos.cpp [hunk 7]
```diff
@@ -88,27 +88,37 @@ void rodeos_db_snapshot::start_block(const get_blocks_result_base& result) {
...
+             ("b", result.this_block->block_num)("f", undo_stack->first_revision())("h", undo_stack->revision()));
+        if (undo_stack->first_revision() >= result.this_block->block_num)
+           throw std::runtime_error("can't switch forks since database doesn't contain revision " +
+                                    std::to_string(result.this_block->block_num - 1));
+        write_session->wipe_cache();
+        while (undo_stack->revision() >= result.this_block->block_num) //
+           undo_stack->undo(true);
+      } else {
+           wlog("can't switch forks at ${b} since undo stack is not enabled. head: ${h}", ("b", result.this_block->block_num) ("h", head));
+           EOS_ASSERT(false, eosio::chain::unlinkable_block_exception, "can't switch forks at ${b} since undo stack is not enabled. head: ${h}", ("b", result.this_block->block_num) ("h", head));
```
[the remark this answers is not in the document]
Comment (linhuang-blockone) [reply libraries/rodeos/rodeos.cpp]: Thanks for the suggestion. Changed to unsupported_feature.

File: programs/rodeos/ship_client.hpp (hunk 5 above)
Comment (linhuang-blockone) [reply programs/rodeos/ship_client.hpp]: Sorry for the typo. Made it not defaulted.

File: programs/rodeos/ship_client.hpp (hunk 6 above)
Comment (linhuang-blockone) [reply programs/rodeos/ship_client.hpp]: Made it not defaulted.

**Review by heifner — APPROVED** [review]
Please update the description of the PR

**Pull request merged**